fanficfare/adapters/adapter_deviantartcom.py

```diff
@@ -0,0 +1,187 @@
+#  -*- coding: utf-8 -*-
+
+# Copyright 2021 FanFicFare team
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+from __future__ import absolute_import
+import logging
+import re
+from datetime import datetime
+# py2 vs py3 transition
+from ..six.moves.urllib import parse as urlparse
+
+from .base_adapter import BaseSiteAdapter, makeDate
+from fanficfare.htmlcleanup import stripHTML
+from .. import exceptions as exceptions
+
+logger = logging.getLogger(__name__)
+
+
+def getClass():
+    return DeviantArtComSiteAdapter
+
+
+class DeviantArtComSiteAdapter(BaseSiteAdapter):
+    def __init__(self, config, url):
+        BaseSiteAdapter.__init__(self, config, url)
+        self.story.setMetadata('siteabbrev', 'deviantartcom')
+
+        self.username = 'NoneGiven'
+        self.password = ''
+        self.is_adult = False
+
+        match = re.match(self.getSiteURLPattern(), url)
+        if not match:
+            raise exceptions.InvalidStoryURL(url, self.getSiteDomain(), self.getSiteExampleURLs())
+
+        story_id = match.group('id')
+        author = match.group('author')
+        self.story.setMetadata('storyId', story_id)
+        self.story.setMetadata('author', author)
+        self.story.setMetadata('authorId', author)
+        self.story.setMetadata('authorUrl', 'https://www.deviantart.com/' + author)
+        self._setURL(url)
+
+    @staticmethod
+    def getSiteDomain():
+        return 'www.deviantart.com'
+
+    @classmethod
+    def getAcceptDomains(cls):
+        return ['www.deviantart.com']
+
+    @classmethod
+    def getProtocol(self):
+        return 'https'
+
+    @classmethod
+    def getSiteExampleURLs(cls):
+        return 'https://%s/<author>/art/<work-name>' % cls.getSiteDomain()
+
+    def getSiteURLPattern(self):
+        return r'https?://www\.deviantart\.com/(?P<author>[^/]+)/art/(?P<id>[^/]+)/?'
+
+    def performLogin(self, url):
```

Comment:
It's worth noting that deviantArt has a bespoke CAPTCHA on it's login screen. I couldn't get it to trigger again, but when I was messing around in a private window I get a CAPTCHA that required me to click and hold a button. I'm not sure what to do about this besides a) figuring out how to reliably trigger to build a solver for it (not sure if that's in my skillset) or b) enable the browser cache feature for deviantArt.

Reply:
I've seen at least one other site that FFF could only download from if you'd visited it in browser recently...

Browser cache feature only supports certain sites--while scanning the cache it only remembers URLs matching patterns it knows to look for.  I haven't generalized it very far as yet.

Reply:
How much work is there to support the cache feature for a new site? I looked at it briefly and it just seemed like a whitelist + cleaning up of URLs to make sure they match (e.g. stripping # parts and query parameters).

To be clear I don't think it will usually be needed since I only ever saw the CAPTCHA once, but if it starts to happen regularly the cache feature would be good to enable for dA.

Reply:
Depends on how much the URLs vary and if they need to be normalized--like ffnet's /s/999/1/ vs /s/999/1/Title.

I agree, something to worry about if and when.

Reply:
I'll take a crack at it for dA just in case. I have yet to see a dA work URL that wasn't just `https://www.deviantart.com/<author>/art/<work id>`, with any anchor or query parameters being unnecessary information.

I'll need to know how it works anyways when I do Patreon support, as far as I can tell FFF's login support won't work with Patreon's 2FA support.

Reply:
Because FFF can run in a CLI, Calibre foreground or Calibre background, I haven't wanted to deal with two factor auth.  

I won't call it impossible, but I haven't had sufficient motivation to work out the details.  

Plus how much different will two factor be for different sites?

Reply:
Hence my desire to sidestep the issue and just use the browser cache.

I'll give it a try later tonight, thanks for the clarifications.